File: backend/app/schemas/payroll.py
from pydantic import BaseModel, Field


# ------------------------------------------------------------------
# Salary Structure & Payroll Calculation Contract
# (field names match the required payload interfaces exactly)
# ------------------------------------------------------------------
class SalaryStructure(BaseModel):
    totalSalary: float
    basicPercentage: float
    basicPay: float
    daPercentage: float
    dearnessAllowance: float
    conveyancePercentage: float
    conveyance: float
    hraPercentage: float
    hra: float
    percentagesSum: float = 100.0
    isValid: bool = True
    validationMessage: str | None = None


class PayrollCalculation(BaseModel):
    totalSalary: float
    basicPay: float
    dearnessAllowance: float
    conveyance: float
    hra: float
    pfWage: float
    pfRate: float
    pfAmount: float
    esiWage: float
    esiRate: float
    esiAmount: float
    professionalTax: float = 0.0
    attendanceDeduction: float = 0.0
    leaveDeduction: float = 0.0
    loanDeduction: float = 0.0
    totalDeductions: float
    netSalary: float


class StructureValidationRequest(BaseModel):
    totalSalary: float = Field(gt=0)
    basicPercentage: float
    daPercentage: float
    conveyancePercentage: float
    hraPercentage: float


class StructureConfig(BaseModel):
    """Dynamic salary-structure percentages (must sum to 100)."""

    basic_percentage: float = 40.0
    da_percentage: float = 20.0
    conveyance_percentage: float = 5.0
    hra_percentage: float = 35.0


class PfConfig(BaseModel):
    """Dynamic PF statutory config.

    wage_base: "BASIC" (only basic pay) | "PF_WAGE" (sum of wage_components).
    """

    enabled: bool = True
    rate: float = 12.0
    wage_base: str = "PF_WAGE"
    wage_components: list[str] = Field(default_factory=lambda: ["BASIC", "DA", "CONVEYANCE"])
    wage_ceiling: float = 15000.0


class EsiConfig(BaseModel):
    """Dynamic ESI statutory config. wage_base: "GROSS" | "ESI_WAGE"."""

    enabled: bool = True
    rate: float = 0.75
    wage_base: str = "GROSS"
    wage_ceiling: float = 21000.0


class PtConfig(BaseModel):
    enabled: bool = False
    amount: float = 200.0


class PayrollSettingsUpdate(BaseModel):
    structure: StructureConfig | None = None
    pf: PfConfig | None = None
    esi: EsiConfig | None = None
    pt: PtConfig | None = None
    standard_working_days: int | None = None
    payroll_cycle_day: int | None = None


class PayrollSettingsOut(BaseModel):
    structure: StructureConfig
    pf: PfConfig
    esi: EsiConfig
    pt: PtConfig
    standard_working_days: int
    payroll_cycle_day: int


class PayrollCalculateRequest(BaseModel):
    employee_id: str | None = None
    month: str = "2026-08"
    bonus: float = 0.0
    reward_earnings: float = 0.0
    attendance_deduction: float = 0.0
    leave_deduction: float = 0.0
    loan_deduction: float = 0.0


class PayrollRecordOut(BaseModel):
    id: str
    employee_id: str
    employee_name: str
    department: str | None = None
    designation: str | None = None
    payroll_month: str
    totalSalary: float
    basicSalary: float
    dearnessAllowance: float
    conveyance: float
    hra: float
    bonus: float
    rewardEarnings: float
    pfWage: float
    pfRate: float
    pfAmount: float
    esiWage: float
    esiRate: float
    esiAmount: float
    professionalTax: float
    attendanceDeduction: float
    leaveDeduction: float
    loanDeduction: float
    totalDeductions: float
    netSalary: float
    workingDays: int
    presentDays: int
    status: str


class PayslipOut(BaseModel):
    company: dict
    record: PayrollRecordOut
    earnings: list[dict]
    deductions: list[dict]
    totals: dict


class PayrollProcessResponse(BaseModel):
    month: str
    processed: int
    records: list[PayrollRecordOut]

File: backend/src/types/payroll.ts
// VRM Enterprise HRMS — Production Payroll Domain Types & Interfaces

export interface SalaryStructureInput {
  monthlySalary: number;
  basicPercentage: number;
  daPercentage: number;
  conveyancePercentage: number;
  hraPercentage: number;
}

export interface ComputedSalaryStructure {
  monthlySalary: number;
  basicSalary: number;
  da: number;
  conveyance: number;
  hra: number;
  fixedGross: number;
  percentagesSum: number;
  isValid: boolean;
}

export interface PfWageComponents {
  basic: boolean;
  da: boolean;
  conveyance: boolean;
  hra: boolean;
  attendance_bonus: boolean;
  overtime: boolean;
  other_earnings: boolean;
  [key: string]: boolean;
}

export interface EsicWageComponents {
  basic: boolean;
  da: boolean;
  conveyance: boolean;
  hra: boolean;
  attendance_bonus: boolean;
  overtime: boolean;
  other_earnings: boolean;
  [key: string]: boolean;
}

export interface PayrollSettingsModel {
  id: string;
  orgKey: string;
  pfEnabled: boolean;
  pfRate: number; // default 12.0
  pfWageCeiling: number; // default 15000.0
  pfWageComponents: PfWageComponents;
  esicEnabled: boolean;
  esicRate: number; // default 0.75
  esicSalaryThreshold: number; // default 21000.0
  esicWageComponents: EsicWageComponents;
  professionalTaxEnabled: boolean;
  professionalTaxAmount: number; // default 200.0
  lopEnabled: boolean;
  attendanceBonusEnabled: boolean;
  overtimeEnabled: boolean;
  standardWorkingDays: number; // default 26
  payrollCycleDay: number; // default 1
}

export interface AttendanceInputData {
  workingDays: number;
  presentDays: number;
  paidLeaveDays: number;
  unpaidLeaveDays: number;
  absentDays: number;
  halfDays: number;
  lopDays: number;
  overtimeHours: number;
}

export interface AdditionalEarningsInput {
  attendanceBonus?: number;
  overtimeHours?: number;
  overtimeRate?: number;
  overtimeAmount?: number;
  bonus?: number;
  incentive?: number;
  commission?: number;
  otherEarnings?: number;
}

export interface AdditionalDeductionsInput {
  advanceRecovery?: number;
  loanRecovery?: number;
  otherDeductions?: number;
}

export interface PayrollCalculationInput {
  employeeId: string;
  monthlySalary: number;
  basicPercentage: number;
  daPercentage: number;
  conveyancePercentage: number;
  hraPercentage: number;
  payrollMonth: number;
  payrollYear: number;
  settings: PayrollSettingsModel;
  attendance: AttendanceInputData;
  earnings?: AdditionalEarningsInput;
  deductions?: AdditionalDeductionsInput;
}

export interface CalculationBreakdownItem {
  name: string;
  category: 'EARNING' | 'DEDUCTION';
  amount: number;
  description?: string;
}

export interface FullPayrollCalculationResult {
  employeeId: string;
  payrollMonth: number;
  payrollYear: number;

  // Fixed Salary Components
  monthlySalary: number;
  basicSalary: number;
  da: number;
  conveyance: number;
  hra: number;
  fixedGross: number;

  // Additional Earnings
  attendanceBonus: number;
  overtimeHours: number;
  overtimeRate: number;
  overtimeAmount: number;
  bonus: number;
  incentive: number;
  commission: number;
  otherEarnings: number;
  grossSalary: number; // Total Gross Earnings

  // Statutory Deductions
  pfBase: number;
  pfRate: number;
  pfAmount: number;

  esicBase: number;
  esicRate: number;
  esicAmount: number;

  professionalTax: number;

  // Attendance & LOP
  workingDays: number;
  presentDays: number;
  lopDays: number;
  dailySalary: number;
  lopAmount: number;

  // Loan & Recovery
  advanceRecovery: number;
  loanRecovery: number;
  otherDeductions: number;

  totalDeductions: number;
  netSalary: number;

  earningsBreakdown: CalculationBreakdownItem[];
  deductionsBreakdown: CalculationBreakdownItem[];
}

export type PayrollRunStatus = 'DRAFT' | 'PROCESSING' | 'PROCESSED' | 'APPROVED' | 'PAID' | 'CANCELLED';

export interface PayrollRunModel {
  id: string;
  payrollMonth: number;
  payrollYear: number;
  status: PayrollRunStatus;
  totalEmployees: number;
  totalGross: number;
  totalDeductions: number;
  totalNet: number;
  processedBy?: string;
  approvedBy?: string;
  processedAt?: string;
  approvedAt?: string;
  paidAt?: string;
  createdAt: string;
  updatedAt: string;
}

export interface PayslipResponseData {
  company: {
    companyName: string;
    legalName: string;
    address: string;
    pan: string;
    gst: string;
  };
  employee: {
    id: string;
    employeeId: string;
    firstName: string;
    lastName: string;
    email: string;
    department: string;
    designation: string;
    bankName?: string;
    accountNumber?: string;
    ifscCode?: string;
  };
  payroll: {
    month: number;
    year: number;
    monthName: string;
    workingDays: number;
    presentDays: number;
    lopDays: number;
    status: string;
  };
  earnings: {
    basicSalary: number;
    da: number;
    conveyance: number;
    hra: number;
    attendanceBonus: number;
    overtimeAmount: number;
    bonus: number;
    incentive: number;
    commission: number;
    otherEarnings: number;
    grossSalary: number;
  };
  deductions: {
    pf: number;
    pfRate: number;
    esic: number;
    esicRate: number;
    professionalTax: number;
    lopAmount: number;
    advanceRecovery: number;
    loanRecovery: number;
    otherDeductions: number;
    totalDeductions: number;
  };
  netSalary: number;
}
